src/main/java/gregtech/common/metatileentities/storage/CraftingRecipeLogic.java

```diff
@@ -0,0 +1,198 @@
+        this.cachedRecipeData = new CachedRecipeData(itemSources, null, inventoryCrafting);
+    }
+
+    public ItemSources getItemSourceList() {
+        return itemSources;
+    }
+
+    public IInventory getCraftingResultInventory() {
+        return craftingResultInventory;
+    }
+
+    public int getItemsCraftedAmount() {
+        return itemsCrafted;
+    }
+
+    public void setItemsCraftedAmount(int itemsCrafted) {
+        this.itemsCrafted = itemsCrafted;
+    }
+
+    public void clearCraftingGrid() {
+        fillCraftinGrid(Collections.emptyMap());
+    }
+
+    public void fillCraftinGrid(Map<Integer, ItemStack> ingredients) {
+        for (int i = 0; i < craftingGrid.getSlots(); i++) {
+            craftingGrid.setStackInSlot(i, ingredients.getOrDefault(i + 1, ItemStack.EMPTY));
+        }
+    }
+
+    private boolean hasCraftingGridUpdated() {
+        boolean craftingGridChanged = false;
+        for (int i = 0; i < craftingGrid.getSlots(); i++) {
+            ItemStackKey oldStack = oldCraftingGrid[i];
+            ItemStack newStack = craftingGrid.getStackInSlot(i);
+            if (oldStack == null) {
+                if (newStack.isEmpty()) {
+                    continue;
+                }
+                oldStack = KeySharedStack.getRegisteredStack(newStack);
+                oldCraftingGrid[i] = oldStack;
+                inventoryCrafting.setInventorySlotContents(i, newStack.copy());
+                craftingGridChanged = true;
+            } else if (newStack.isEmpty()) {
+                oldCraftingGrid[i] = null;
+                inventoryCrafting.setInventorySlotContents(i, ItemStack.EMPTY);
+                craftingGridChanged = true;
+            } else if (!ItemStack.areItemsEqual(oldStack.getItemStackRaw(), newStack) ||
+                    !ItemStack.areItemStackTagsEqual(oldStack.getItemStackRaw(), newStack)) {
+                oldCraftingGrid[i] = KeySharedStack.getRegisteredStack(newStack);
+                inventoryCrafting.setInventorySlotContents(i, newStack.copy());
+                craftingGridChanged = true;
+            }
+        }
+        return craftingGridChanged;
+    }
+
+    public void performRecipe(EntityPlayer player) {
+        if (!isRecipeValid()) {
+            return;
+        }
+        if (!cachedRecipeData.consumeRecipeItems()) {
+            return;
+        }
+        ForgeHooks.setCraftingPlayer(player);
+        NonNullList<ItemStack> remainingItems = cachedRecipe.getRemainingItems(inventoryCrafting); // todo right here is where tools get damaged (in UI)
+        ForgeHooks.setCraftingPlayer(null);
+        for (int i = 0; i < remainingItems.size(); i++) {
+            ItemStack itemStack = remainingItems.get(i);
+            if (itemStack.isEmpty()) {
+                continue;
+            }
+            ItemStackKey stackKey = KeySharedStack.getRegisteredStack(itemStack);
+            inventoryCrafting.setInventorySlotContents(i, itemStack);
```

Comment:
Is this line necessary? We have our own storage for the crafting grid, and the crafting grid is all ghost items anyway, so I am not sure if this does anything or is required.

Reply:
the inventory represented by `craftingGrid`is the ghost one, shown to the player.
the inventory that holds the items that actually are gonna be used in the recipe is `inventoryCrafting`, so we update it with the returned containerItem so it reuses the last damaged tool instead of iterating the inventory